Script compiler refactoring

fernewelten opened the pull request:
**Summary of additions to the language:**

Structs can contain variables that are other structs  In this case, managed structs can only be included via pointers; non-managed structs may not be included via pointers.
```
struct Pair { 
    int X, Y; 
};
managed struct Trio {
    int X, Y, Z; 
};

struct TalkingPositions { 
    Pair PosOfChar1, PosOfChar2; // these components can't be Pair * because Pair isn't managed
}; 
struct FloatingHead { 
    Trio *  HeadPos;     // HeadPos must be pointer because Trio is managed;
}; 
``` 
Can now have dynamic arrays of elements that contain dynamic arrays. 

Functions can now  be defined in any sequence as long as they are in the same compilation unit. So you can reference a function first and define it afterwards. This makes recursive algorithms much easier to code.
```
int  Func1(int a)
{
    return 2 * Func2(a);    // uses Func2()
}

int Func2(int a)     // defines Func2() and is below its first use -- this used to be illegal.
{ … }
```

Forward declarations of functions don't need the "import" keyword if you don't need the functions to be importable elsewhere. Basically, say what you mean: If a function is meant to be importable, use "import"; if it isn't, don't use it. 

```
import int foo (int a, int b = 5, int c = 7); // okay, declares a function that can be imported elsewhere. Header files have lots of those.

struct s
{
    int a, b;
    function foo(int a); // declares the function s::foo() but doesn't make it importable -- so no "import" keyword
};
```


Defaults for function parameters can be defined in any function header no matter whether the function body follows; however, for all the function headers that name defaults at all, they must name _all_ the defaults and these must be exactly the same. 
```
import int foo (int a, int b = 5, int c = 7); // okay, declares some parameter defaults for that function

int foo (int a, int b = 5, int c = 7) // okay
{ 
   ... 
}
int foo (int a, int b, int c); // okay, doesn't name ANY defaults -- can have any number of those lines
int foo (int a, int b = 5, int c = 7 ); // okay, names ALL the defaults that exist  -- can have any number of those lines
int foo (int a, int b, int c = 7) // not okay, names the default for c but not for b
int foo (int a, int b = 6, int c = 7) // not okay, b has a different default than in other declarations of foo.
int foo (int a, int b, int c) // would be okay if there wasn't another function foo with a body already 
{
   ...
}
```

Ternaries: `int foo = (x > 5) ? 1 : 2;` -- foo is set to `1` if `(x > 5)`, to `2` otherwise.

Can leave out the second part of a ternary so that "?" and ":" go right next to each other. For instance,
```int foo = bar ?: 15; ``` 
In this case, if `bar` is  0 or `null` then the result is `15` else the result is `bar`. Handy for specifying a default in case a variable turns out to be `null`.  (This is the same functionality as the `??` operator in C#.)

As in C or C++, can define a struct and global variables in the same statement . For instance, the following will define a vartype `Pair` and a variable `Pair1` that has this new vartype. 
```
struct Pair 
{ 
    int X, Y; 
} Pair1;
``` 


When a variable is declared as a `managed` type, then the compiler _knows_ that it must be a pointer variable because non-pointers to managed types are illegal. (See below for one big exception to this rule.) If you forget the "*" in these cases, then the compiler will happily imply it and carry on compiling. For instance, in the example below, `Var1`, `Var2`, and `W` will be implied to have the type `Weapon *` because `Weapon`  is a `managed` struct.
```
managed struct Weapon
{
        int A;
} Var1;
     
Weapon Var2;
     
function Foo(Weapon W)
{ … }
```
In practice, this means that you can get away with omitting nearly all the pointer symbols in your AGS code.

The big exception is `import` declarations. In import declarations, it _is_ legal to declare `managed struct`s  that aren't pointers, and automatically generated AGS code will actually do that, so the compiler does _not_ imply any `*` for things that are declared `import`. Any pointers must be explicitly written out there.

Multi-dimensional non-dynamic arrays: `int a[5, 3];` or equivalently, `int a[5][3];`. You can have any positive number of dimensions `int a[2][3][5][7][11];`.

String and character literals accept any `\` codes that C or C++ accept, e.g. `\0` or `\n`. (Didn't change the customary handling of `[`.)

Sequences such as `a.b.c[17 * d].e[f]` may now be arbitrarily long. 

Attributes / properties: The getter and setter functions of attributes can now be defined in the same compilation unit as the respective attributes.



**Summary of changes to the compiler code:**

Touched up the parser code: Tried to enhance comprehensibility by:
- putting in functions
- seriously reducing nesting depth
- adding in comments
- renaming some variables and functions to be as apt as I could make them
- converting the code to the AGS formatting/indentation guidelines on the AGS repository wiki

Wrote googletests that cover > 80 % of the code lines

Converted the scanner and parser to C++ classes in the   `AGS` namespace.

Converted the arrays that manage code nesting to a proper class in the`AGS` namespace.

Converted to C++ strings (std::string). The scanner and parser  don't use any C string functions (strcpy etc.) at all, or C strings apart from string literals.

Found an "old" symboltable structure lying around in script/cc_symboldef.h that had been nearly, but not completely superseded by the "new" symboltable structure in script/symboltable.h. As far as I could ascertain, the parser was the last one to use the "old" symboltable. Rewrote these code places to use the "new" symboltable structure, so the new parser doesn't use it any more, either.

Revamped the variable access code,

Added some typedefs in the ags namespace that helped me check my code. Most notably, `AGS::Symbol` for integers that point into the symbol table.

Converted lots of bit yoga into inline functions `FlagIsSet()` and `SetFlag()` for better readability. Then encapsuled a lot of them into dedicated inline functions (such as `IsStructMember()`) and moved them to the symbol table class.

Moved the symbol table class into AGS namespace. Relocated all functions that work with symbols into this symbol table class, e.g., `IsAnyTypeOfString()`, `IsDynarray()` etc. 

Changed the `#define`s for symbol types into an `enum` for better debugging. 

Changed some error messages to be more descriptive. 

Many error messages that point out a discrepancy of current use to former code will now name the exact location of the former code that clashes with the current use. (`This function has 3 parameters. This clashes with another definition. See globalscript.ash, line 332`);

The scanner checks that all opening symbols have matching closing symbols. Code like `int a[9)` or `for(i = 0; i < 5;` _(end of code)_ won't even pass the scanner. 

The scanner accepts and converts all the backslash sequences that C or C++ accept (`\0`, `\n`etc.). (This processing does not change the customary handling of '['). All such string literal handling has been relocated to the scanner.

In the original compiler, functions that defined long static arrays of dynamic pointers would yield very long Bytecode sequences. For instance, 
```
int foo()
{
    Object *bar[10000];
}
```
would generate 10000 individual Bytecode statements to release each of the array element pointers. Found a way to release the pointers in a loop instead. So in this case the compiler only generates about a half-dozen of Bytecode statements.



**Caveats:**

- string buffers (`string foo`) are barely tested.

rofl0r commented:
this looks as if you just copied your entire work over this branch in a single huge commit o.0
(and including 5 MB binary files). i would suspect that it would be much easier to review if you could preserve the initial commits (this can be done e.g. by extracting the old commits from the other branch with git format-patch and them apply them to the new branch, either automatically (if possible) with git am or manually.

i personally would never merge a 16.000 changed-lines-patch, but these are just my 2 shekels...

ivan-mogilko commented:
> (and including 5 MB binary files)

Right, so there is another unnecessary file that should not be in .gitignore: ".vs/slnx.sqlite" (looks like a temp output from MSVS?)

Regarding the commits in general, the basic guidelines of creating these are:
* One subtask per commit. For example, if you are merely fixing codestyle in a file and do nothing else then make a separate commit for that. If you are also fixing some error in there then it's better to make separate commit before or after fixing codestyle. I swear, this really helps if someone has to trace history of changes back for any reason (like finding when and why program behavior has changed).
* Commits should have meaningful names. For example: "Applied codestyle", "Refactored parser code", "Fixed error (with error description".
"Port from ags4" name does not explain anything. We know what this means right now because we know that you first worked in the "ags4" branch, but imagine someone else investigating these changes several months later.

I don't know how easy that would be for you to rearrange your changes into distinct commits, but I ask you to give it a try, because that really can make a huge difference when reviewing the new code as well as when searching for new bugs.

Note that if you change commits in your work branch you may then force-push it again and github will automatically replace the contents of pull request.

fernewelten commented:
Still about 3 times the change lines I was expecting - all the script files _except_ for cs_parser.cpp ought to be nearly unchanged. As far as I am able to see, most of the spurious change lines seem to have been introduced by Visual Studio trying to reformat code (for instance, inserting a space after a comma). I'll try to get these out of the picture.

rofl0r commented on Compiler/script/cc_symboltable.cpp:
```diff
@@ -140,107 +150,112 @@ void symbolTable::reset() {
+        return nameGenCache[idx];
+    }
 
-	std::size_t actualIdx = idx & STYPE_MASK;
```
there are tools like gitk that let you look at your commits locally, so you don't have to push to github to see that you introduced gratuitous whitespace changes.

ivan-mogilko replied:
I think these could be intentional, fernewelten mentioned he was adjusting codestyle to match our convention.

ivan-mogilko commented:
@fernewelten, could you tell what do these numbers in commit names stand for?

ivan-mogilko commented:
Also, I did not realize this at first, but you've merged the older version of this branch into new version, thus bringing these 3 big original commits into pr again...

rofl0r commented on Compiler/script/cs_parser.h:
```diff
@@ -70,14 +57,14 @@ class OpenCloseMatcher
 
 private:
-    struct OpenInfo
+    struct OpenInfo_t
```
> POSIX ... defines a lot of extra type names ending in '_t', and reserves the suffix for the implementation. That means that if you are working on POSIX-related systems, defining your own type names with the convention is ill-advised

fernewelten replied:
Removed.

ivan-mogilko commented:
@fernewelten  This is my oversight but I just noticed that our code convention draft contains naming rules that actually conflict with the naming convention we use in practice in the C++ parts of this project.

Following paragraph:
<pre>
For most cases, "upper" CamelCase notation (first and medial letters capitalized) is the preferred naming convention. Namespaces, classes, and enums should always be named in upper CamelCase. Publicly exposed data (class members, methods, globals, etc.) should also follow CamelCase notation. Private members and methods, and other private data should be named using "lower" camelCase notation (first letter lower case, medial letters capitalized).
</pre>

Should read:
<pre>
... Private fields (***not methods!***) should be named using "lower" camelCase notation (first letter lower case, medial letters capitalized) ***preceded with a single underscore symbol ('_')***.
</pre>

Probably we had this changed after the text was written.

I will scan through the page again to see if there is other false information there.

ivan-mogilko commented:
Besides above, looking at your recent changes:

* Namespaces should have camel-case format. In case of namespace "ags" it should be "AGS" (since its an abbreviation)
* Class and struct names should not have this "_t" postfix, this is reserved for primitive types.

fernewelten commented:
Loads of comments; many thanks for all the feedback!

I'm a big fan of small steps, too, but in the present case I'm afraid I haven't managed to make the step more “atomic” than it is. For all intents and purposes, this has turned into a big-lump rewrite, although it does mimic the former logic almost completely from a very high bird's eye view. I've tried very hard to keep in every single thougt. 

That's one of the reasons I'm aiming to cover as near as 100 % of the code lines as I can get with googletests that check byte-for-byte that the new code emits exactly the same bytes as the old code. 

Gurok has pointed me to documentation to attributes and stringstructs. With that help, I'll check the remaining coding, adding googletests as I go along. Until then, I don't trust the corresponding pieces of code.

> * Namespaces should have camel-case format. In case of namespace "ags" it should be "AGS" (all caps since its an abbreviation)
Haven't touched that name yet; it's from the pre-existing code. Happy to change it to "AGS".

> Publicly exposed data (class members, methods, globals, etc.) as well as private methods should also follow CamelCase notation. 
Great! Will be glad to change that.

> rationale for removing this comment ?
I'm currently rewriting and extending it a bit offline (in a "proper" editor) and will put it back in afterwards, probably today. No content will get lost. Part of the information can now be copied nearer to where the specific logic happens. As I read it, the comment at the top of the file is meant as a “bird's eye” summary, and I hope to come up with some more thought for that.

> why not prefix these variables with "the" too ?
You're right, that would be more consistent. The "the" have crept in to avoid name clashes and to have at least one "hump" for camelcasing.

> @fernewelten, could you tell what do these numbers in commit names stand for?
Those were internal bookkeeping; just ignore them; I've discontinued them already.

fernewelten commented on Compiler/script/cs_parser.h:
```diff
@@ -126,14 +113,13 @@ class Scanner
-    std::string LastError;
+    std::istringstream inputStream;
+    std::size_t theLineno;
+    ccInternalList *tokenList;
```
With the naming convention for private vars, the "the" could disappear, so consistency is reached.

fernewelten commented on Compiler/script/cs_parser.h:
```diff
@@ -1,21 +1,7 @@
 'C'-style script compiler development file. (c) 2000,2001 Chris Jones
 
-SCOM is a script compiler for the 'C' language. The current version
-implements:
```
New comments in place.

ivan-mogilko commented:
Wanted to mention only fyi, I started moving coding convention to repository's wiki, it may be found here: https://github.com/adventuregamestudio/ags/wiki/AGS-Coding-Conventions-(Cpp)

It's only about a half of previous content yet, and I am trying to review and fix some things there as the old draft have became partially outdated.

ivan-mogilko commented:
@fernewelten your latest commit contains unnecessary visual studio output again. To prevent these, I think, you may add ".vs" to gitignore.

fernewelten commented:
I begin to see where the problem is: I've just updated VS. Last time those spurious files showed up, I'd updated VS, too. Microsoft might currently be very “inventive” with creating new kinds of temporary files.

rofl0r commented:
in the last commit a lot of bogus binary files were added.
```
BIN +6.5 KB Editor/AGS.CScript.Compiler/.vs/AGS.CScript.Compiler/v15/.suo
0 Editor/AGS.CScript.Compiler/.vs/AGS.CScript.Compiler/v15/Server/sqlite3/db.lock
BIN +504 KB Editor/AGS.CScript.Compiler/.vs/AGS.CScript.Compiler/v15/Server/sqlite3/storage.ide
BIN +7 KB Editor/AGS.Editor/.vs/AGSEditor/v15/.suo
0 Editor/AGS.Editor/.vs/AGSEditor/v15/Server/sqlite3/db.lock
BIN +4 KB Editor/AGS.Editor/.vs/AGSEditor/v15/Server/sqlite3/storage.ide
BIN +32 KB Editor/AGS.Editor/.vs/AGSEditor/v15/Server/sqlite3/storage.ide-shm
BIN +1.08 MB Editor/AGS.Editor/.vs/AGSEditor/v15/Server/sqlite3/storage.ide-wal 
```
these should be removed with an interactive rebase.
the mistake you're making is that you use `git add -a` to add **all** files that were added or changed. that is a mistake. you should look before every commit which files were changed with `git status` or `git diff` (assuming you don't want to use `gitk` or a similar gui that visualizes changes and commits) and add them manually.
if only changes were made that you actually want to check in, `git add -u` adds only (and all) changed files that were previously checked in.

properly reviewing stuff before commiting makes the difference between a high-quality repo and a junkyard.

fernewelten commented:
> in the last commit a lot of bogus binary files were added.

yes, this has been noted (https://github.com/adventuregamestudio/ags/pull/585#issuecomment-457142712), and I'll probably be able to see to it this Sunday evening  –I'm nearly completey offline for the week-end. 

I have been pointed to Visual Studio as being **one of the main compilers** of this project -- seeing that the .vs directory has been its main repository for  internal files ever since Visual Studio 2015 (cf. above, https://github.com/adventuregamestudio/ags/pull/585#issuecomment-457155929), 

how come that this directory hasn't been added to .gitignore **about three years ago**? I was using the .gitignore that was already in the project.

Who **is** actually using VS?

ivan-mogilko commented:
I was using Visual Studio, but apparently mine did not create the ".vs" folders in these particular locations. I cannot tell why, because I don't know under which conditions it does so.

rofl0r commented:
> how come that this directory hasn't been added to .gitignore about three years ago?

that's no excuse for not looking at the stuff you're about to commit and just blindly adding **everything**. maybe you should read more than the first sentence in my last comment.

jdiperla commented:
I think typing things sound harsher that the user meant to make it seem. I
often come across this way because of language limitations.. Dont get me
wrong, I too actually mean to sound harsh from time to time. But in this
case, I think it sounded worse than it was meant to come off.

-------------
Giuseppe DiPerla
Deep River, CT Congregation
Phone: 860-518-1500
Email: JDiPerla@GMail.com ; Giuseppe@DiPerla.com
Google Voice: 860-615-9428


On Fri, Jan 25, 2019 at 9:43 AM rofl0r <notifications@github.com> wrote:

> how come that this directory hasn't been added to .gitignore about three
> years ago?
>
> that's no excuse for not looking at the stuff you're about to commit and
> just blindly adding *everything*. maybe you should read more than the
> first sentence in my last comment.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adventuregamestudio/ags/pull/585#issuecomment-457594226>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1wh0_Yuonf-3f2B7JU30IjXhXxIugOks5vGxgagaJpZM4aAElc>
> .
>

ivan-mogilko commented:
@fernewelten please put a notice when you are done adding refactor commits so that I could review all thing in entirety again.

I'm afraid there will be a need to do another rebase to clean the branch up, because these 3 huge commits ("Port from ags4" etc) were not removed but merged instead (merge commit 896c095f0e4e79b496b71751129b2a97f0bee00b); and also this will be a chance to get rid of unnecessary files and do proper .gitignore edit.

ivan-mogilko commented:
@fernewelten, I apologize for more delay, I was still not sure if you've finished making commits for this pull request.

There are couple of further notes I want to make.

---

**1)** First of all, I pushed some fixes to compiler tests in the master branch just now because the compilation was broken after certain changes in shared code. After my fixes your PR branch now has conflicts with our master. Besides trivial ones in .gitignore and MSVS projects, in compiler test library **last_seen_cc_error** is no longer a char* pointer, but a function that returns const char* pointer from a std::string. I did a quick test and following adjustment will make your code work again (it's possible to do with "replace all"):

| old code | new code | note |
|---|---|---|
| last_seen_cc_error = 0; | clear_error(); ||
| last_seen_cc_error); | last_seen_cc_error());| usually this is last argument in a function |
| std::string err = last_seen_cc_error; | std::string err = last_seen_cc_error(); | OR std::string err = last_cc_error_buf; |

---

**2)** I took your branch and tried to compile one of my existing game projects, but there was a compilation error in built-in header:

`Failed to save room room305.crm; details below
_BuiltInScriptHeader.ash(1123): Error (line 1123): Expected a variable type instead of 'bool'`

The referred line is this part of AGS API:

<pre>builtin managed struct ListBox extends GUIControl {
	/// Adds a new item to the bottom of the list with the specified text.
	import bool AddItem(const string text);
</pre>

I'd like to keep testing your new compiler code with my project because it has lots of various code in it. It's actually open-source, but I forgot to load the code into some repository, which I'll do after properly updating it to the newer version of AGS.

---

Besides that, my suggestion as earlier is to make a clean rebase on top of the current master, because your branch is based off a pretty outdated version now, and at the same time it will be a chance to clear redundant commits (like 896c095) and unnecessary files in commits.

fernewelten commented:
> I did a quick test and following adjustment will make your code work again (it's possible to do with "replace all"):


Okay, I'll make a dedicated commit for that. 

> Failed to save room room305.crm; details below _BuiltInScriptHeader.ash(1123): Error (line 1123): Expected a variable type instead of 'bool'

This is bound to be the code that processes `enum`s because the parser itself doesn't know `bool`. Instead, in Editor\AGS.Editor\Resources\agsdefns.sh, line 82, there is:

```
enum bool {
  false = 0,
  true = 1
};
```
So probably the parser doesn't deduce that "`bool`" is a kind of `int` after reading that declaration.

I'll get onto that. 

> I'd like to keep testing your new compiler code with my project because it has lots of various code in it. 

IMO, this is a great idea to uncover errors that the parser might still have.

ivan-mogilko commented:
> > I did a quick test and following adjustment will make your code work again (it's possible to do with "replace all"):

> Okay, I'll make a dedicated commit for that.

I really suggest a rebase instead... over time the conflicts like this may accumulate.

fernewelten commented:
I'm working off the solution `ags\Solutions\Compile.Lib.sln`

Early in the rebasing process, I'm getting a bunch of unresolved symbols:
```
2>Compiler_d.lib(cc_error.obj) : error LNK2019: unresolved external symbol "public: __thiscall AGS::Common::String::String(void)" (??0String@Common@AGS@@QAE@XZ) referenced in function "void __cdecl `dynamic initializer for 'ccErrorCallStack''(void)" (??__EccErrorCallStack@@YAXXZ)
2>Compiler_d.lib(cc_error.obj) : error LNK2019: unresolved external symbol "public: __thiscall AGS::Common::String::~String(void)" (??1String@Common@AGS@@QAE@XZ) referenced in function "public: __thiscall std::pair<class AGS::Common::String,class AGS::Common::String>::~pair<class AGS::Common::String,class AGS::Common::String>(void)" (??1?$pair@VString@Common@AGS@@V123@@std@@QAE@XZ)
2>Compiler_d.lib(cc_error.obj) : error LNK2019: unresolved external symbol "public: class AGS::Common::String & __thiscall AGS::Common::String::operator=(class AGS::Common::String const &)" (??4String@Common@AGS@@QAEAAV012@ABV012@@Z) referenced in function "void __cdecl cc_error(char const *,...)" (?cc_error@@YAXPBDZZ)
2>Compiler_d.lib(cc_error.obj) : error LNK2019: unresolved external symbol "public: class AGS::Common::String & __thiscall AGS::Common::String::operator=(char const *)" (??4String@Common@AGS@@QAEAAV012@PBD@Z) referenced in function "void __cdecl cc_error(char const *,...)" (?cc_error@@YAXPBDZZ)
2>Compiler_d.lib(cc_error.obj) : error LNK2019: unresolved external symbol "struct std::pair<class AGS::Common::String,class AGS::Common::String> __cdecl cc_error_at_line(char const *)" (?cc_error_at_line@@YA?AU?$pair@VString@Common@AGS@@V123@@std@@PBD@Z) referenced in function "void __cdecl cc_error(char const *,...)" (?cc_error@@YAXPBDZZ)
2>Compiler_d.lib(cc_error.obj) : error LNK2019: unresolved external symbol "class AGS::Common::String __cdecl cc_error_without_line(char const *)" (?cc_error_without_line@@YA?AVString@Common@AGS@@PBD@Z) referenced in function "void __cdecl cc_error(char const *,...)" (?cc_error@@YAXPBDZZ)
```
These seem to be caused by `\Common\script\cc_error.cpp`, which I haven't touched. 
Half of the lines go away by including `\Common\script\cc_script.h` and `\Common\script\cc_script.cpp` into the solution. However, the existing `error.cpp` also depends on the following externals that seem to have disappeared from the codebase:

```
// Returns full script error message and callstack (if possible)
extern std::pair<String, String> cc_error_at_line(const char *error_msg);
// Returns script error message without location or callstack
extern String cc_error_without_line(const char *error_msg);
```
What's the replacement?

ivan-mogilko commented:
Right, I forgot to mention, you need to include "Common/util/string.cpp", that should be enough.

ivan-mogilko commented:
> Early in the rebasing process, I'm getting a bunch of unresolved symbols:
> However, the existing `error.cpp` also depends on the following externals that seem to have disappeared from the codebase

Hmm, functions cc_error_at_line and cc_error_without_line are only declared in Compiler.lib, but they have separate implementations in the Editor, Engine and Compiler.Lib.Test, depending on what you are linking Compiler.Lib to.

But recently I was fixing this problem in Compiler.Test.Lib, see commit: 435d11843ecbb0e160feb5d2b9692d42829a8f56 and it is merged into master with commit 846a82327fea138676bc14da7ecc4b1adeba84da.

It should be in master branch already. Could you tell which commit in master are you rebasing on?

fernewelten commented:
> PS. I was thinking, maybe I could help if I recreate your branch by cherry-picking your commits into new branch in my own repository and which you could checkout for yourself later? Since I already tested to merge your work into current master, maybe it will be easier to do for me.

Thanks a lot, but that wouldn't be the sustainable, long term solution, would it? I'm currently assuming that my setup must be incorrect in some way: 

As an experiment, I included `Editor\Native\script_agsnative.cpp` into the solution, and the code compiled again. So I'm assuming that my linker doesn't find some library-type or assembly-type file it **should** find – or else that it generates its output in a directory it **shouldn't** generate to. The sensible thing to do would be to correct the setup.

I'm afraid that this “Visual” setup is much less intuitive to me than, e.g., an Autoconf/Makefile setup would be --in the latter, I'd have succeeded long ago in checking and fixing library directories. And I'm finding the Internet to be surprisingly unhelpful for specific VS matters whilst giving loads of irrelevant and/or outdated information at the same time.

ivan-mogilko commented:
> Thanks a lot, but that wouldn't be the sustainable, long term solution, would it?

To clarify just in case, I meant, you would take over the new branch and continue this PR with it (force pushing with same name).
Well, anyway, I could do that just for a test and to see if something comes up when I have spare time.

> I'm currently assuming that my setup must be incorrect in some way

As of the "master" branch, for Compiler.Test solution these functions are implemented in cs_parser_test.cpp. I was assuming it did not go anywhere in your branch...
If you build Engine or Editor these are implemented in script_engine and script_agsnative, iirc.

fernewelten commented:
So, let's hope that this time all the right files have come over and no wrong ones. 
This is rebased off ab6623bc “Merge branch 'ags3'” and still passes the 109 googletests.

ivan-mogilko commented:
Oh, it looks like you've kept those 3 giant commits I kept mentioning about ("port from ags4" and 2 following ones), and put everything else on top of them. Compared to the previous branch, some of the latter commits lost their diff content, probably because it was already done in one of those 3.

ivan-mogilko commented:
I have pushed the *old version* of your branch to my personal repository as a backup:
https://github.com/ivan-mogilko/ags-refactoring/commits/Script_compiler_refactoring

To illustrate what I meant by above comment, for example compare commit called "Automatic reformatting in cc_compiledscript.cpp, move yank_chunk(), write_chunk(), start on NestingStack 30"

The old one: https://github.com/ivan-mogilko/ags-refactoring/commit/86054e4f57d7eb455384c540278081266d57ab62
The new one in the rebased branch: https://github.com/adventuregamestudio/ags/pull/585/commits/44a0b23f87efa34950e055062ff2410aee5365f9

ivan-mogilko commented:
Hmm, I made a very quick test trying to rebase the previous variant of a branch, but found that at least several first commits do not compile (do not contain full logical changeset). So I really don't know what way is better.

fernewelten commented:
_All_ the rebased commits compiled before I checked them in. In the original commits, some extraneous files crept in; I tried very hard to not have that happening this time by .gitignoring them.

ivan-mogilko commented:
Sorry, I think we have a very bad case of miscommunication going on in regards to this pull request. I am no longer sure if we get each other intentions clearly.

I was going to try explaining what I was asking about one more time, but then I thought this may start new round of misunderstanding and waste everyones time.
Besides, since this is practically a full rewrite of whole parser and it already took a lot of effort, I guess it's better to just treat it as one big change of code and move on.

---

I'll put my explanation under this "spoiler", only as optional information.

<details>
Originally one problem was that your PR consisted of three huge commits which changed a lot of code at once. After we noted that, you have split them up into multiple commits which did about same changes in summary. But then instead of replacing original 3-commit branch with the newer one you somehow merged two branches, - one containing 3 big commits and another containing same changes but in multiple commits, - and pull request became looking weird (which I noted several times in this discussion).

Visually, previously PR branch looked like this:
<pre>
X-----A -> B -> C --------------- merge commit ---------- the rest of refactoring-->
 \-----a-a-a-a-b-b-b-b-b-c-c-c-c---/
</pre>

Now it looks like this:
<pre>
X-----A -> B -> C -a-a-a-a-b-b-b-b-b-c-c-c-c---- the rest of refactoring -->
</pre>

Since a bunch of small commits practically duplicated first 3 large ones, in the new branch they lost most of their content and no longer correspond even to their names.

So for example, a commit is called "Calculation of jump offsets". In old branch was it actually doing something (https://github.com/ivan-mogilko/ags-refactoring/commit/bfc8c7fea5fbcfd682ae59076d4003e7a63d63a6)
in this new branch it does nothing except changing 1 line in the test: https://github.com/adventuregamestudio/ags/commit/6e8f5ff7f1d5eac087ec10fb605ed1ce9243a716

I was simply trying to bring your attention to this, because I'd like to know if your intentions were really to have something like this, or not.
</details>

fernewelten commented:
As concerns the .gitignore, it seems it isn't enough to just exclude "`.vs`" or "`.vs/`". As far as I could see, this only prevents the **directory** itself from getting into the changes, not the files that are **inside** the directory, Git  showed that it disregarded the `.vs` itself, but wouldn't let me finish the commit because some `.vs/soandso` files had locally changed and weren't staged.
In the end, I tried "`**/.vs/**`", which seemed to work. I put it next to the other Windows specific commands.

Also, I'm using `OpenCppCoverage`, a Public Domain tool to show me just what codelines are covered by my Googletests. All the lines that the program flow runs through are higlighted in green; all the lines that aren't are highlighted in red so I know that they can't have been tested yet. However, it seems that the program writes out a logfile, `LastCoverageResults.log`. The program isn't specifically VS, so it does __not__ park its log file in the `.vs/` directory, so the existing .gitignore rules didn't cover it. I put `**/LastCoverageResults.log`  in and hope that this will keep this file out of the commits no matter **where** it is written.

ivan-mogilko commented:
About .gitignore, following entries were deleted by merge mistake, I think:
https://github.com/adventuregamestudio/ags/pull/585/commits/fecba24d419d355139d9d27139a196fae83a73ee#diff-a084b794bc0759e7a6b77810e01874f2L143

<pre>
# macOS system files
.DS_Store
.AppleDouble
.LSOverride
._*


# Windows system files
thumbs.db
desktop.ini


# recommended CMake build directory
build*/**
</pre>

fernewelten commented:
> I took your branch and tried to compile one of my existing game projects, but there was a compilation error in built-in header:
>> Failed to save room room305.crm; details below _BuiltInScriptHeader.ash(1123): Error (line 1123): Expected a variable type instead of 'bool'
The referred line is this part of AGS API:
https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Editor/Resources/agsdefns.sh

That turned out to be a strange edge case for the tokenizer, not the parser. The tokenizer didn't work properly whenever a `struct` was defined with an `extends` clause and that `struct` had a component with a non-primitive type. I think I have fixed that now.

BTW., is there a way to run the preprocessor separately? The compiler can only grok preprocessed code, so I can't feed the file `Editor/AGS.Editor/Resources/agsdefns.sh` to the compiler directly. If there isn't, I'll try and cobble together a stand-alone C# program that uses the existing preprocessor classes.

fernewelten commented:
> About .gitignore, following entries were deleted by merge mistake, I think
I've pushed through a commit that restores the .gitignore files that were missing.

ivan-mogilko commented:
Ohhh, I am awfully sorry!! Probably I missed there was a fix to previously reported issue. I kept getting distracted by my own tasks all the time.
I promise to pay closer attention to this from now on.

I tried again with the newer branch. The previous bug was fixed, but I found new ones:

1. Declaring a struct member function of name identical to a standard global function causes "already defined" error (I think we had this bug in the past with the existing compiler but it was fixed).

For some reason it works only if name matches **builtin function**, and I don't know why!

Following example causes the error:
<pre>
struct MyStruct
{
    import int GetTextWidth(String s);
}
</pre>

At the same time following works, at least for me:
<pre>
import int MyFunc(String s);

struct MyStruct
{
    import int MyFunc(String s);
}
</pre>

2. Sometimes reported errors have corrupted function names. For example if I do this:
<pre>
import int MyFunc(String s);
import int MyFunc(String s);
</pre>

It correctly reports "already defined" error, but the symbol name consists of random characters or is empty string.

3. Variadic functions and/or their calls are parsed incorrectly. Calling String.Format with multiple arguments results in "Expected just 1 parameters but found N".

4. Comparing dynamic array to null is forbidden for some reason:
<pre>
int xxx[] = new int[10];
if (xxx == null)
{
}
</pre>
This results in "The operator cannot be applied to values of these types".

5. Assigning pointer to the array of pointers does not work.
<pre>
DynamicSprite *sprites[] = new DynamicSprite[10];
DynamicSprite *spr = null;
sprites[0] = spr;
</pre>
Error: Type mismatch: cannot convert 'DynamicSprite*' to 'DynamicSprite*[]'

6. There's a bug in ParseVardecl_CheckIllegalCombis which causes printf exception, because it uses formatting string without actual arguments:
<pre>
cc_error("Symbol '%s' is already defined");
</pre>



I want to upload my game's sources tomorrow, after I adjust them for the master branch (ags4), it has a lot of various scripts and may be good for testing.

ivan-mogilko commented:
@fernewelten , here's the game I was testing your compiler with: https://github.com/ivan-mogilko/ags-lastfurious/tree/dev--ags350
(use "dev--ags350" branch, it is adjusted to compile in ags4 branch)

fernewelten commented:
So, by now I found a way to get original AGS code preprocessed and assembled in exactly the same way as if it had come through the Editor. This makes it possible to take a real-world game and test whether the new compiler will emit the same Bytecode for it as the old one did.

I've already done this with one game, and that has resulted in a lot of improvements. When I'm done testing @ivan-mogilko 's game, too, I'll open the flood gates again.

fernewelten commented:
@ivan-mogilko , I can't seem to open the game. I'm using the dev--ags350 branch. When I open `Game.agf`, the studio shows the following error message: 
![errmsg](https://user-images.githubusercontent.com/41213694/54890899-c3a11a00-4eab-11e9-9398-3ce1e2fe21c9.png)

Unfortunately, the Editor talks German in part. Here are translations:
* German: "Der angeforderte Wert "v3507" konnte nicht gefunden werden" 
* English "The requested value "v3507" could not be found".

In the error details:
* German: "bei"
* English: "at"
* German: "Zeile"
* English: "line"

ivan-mogilko commented:
>  I can't seem to open the game. I'm using the dev--ags350 branch. When I open Game.agf, the studio shows the following error message:

Do you build Editor yourself, or use particular release? I think it should be opened by latest master branch too. But I will test again.

sonneveld commented:
Considering AGS3 is still in development, and changes done in that branch flow into AGS4, would you be willing to rebase this off the ags3 branch instead?

fernewelten commented:
> Considering AGS3 is still in development, and changes done in that branch flow into AGS4, would you be willing to rebase this off the ags3 branch instead?

I'm careful in keeping all the changes to the compiler in such a way that existing programs won't break.  Currently the compiler even generates byte-identical code to the non-rewritten one. It's self-contained and uses "vanilla C++/C". So yes, it can be moved into AGS3 I think.

Before rebasing it just yet, I'd like to finish something I'm actively working on at the moment, however.

fernewelten commented:
In my setup, the solutions AGS.Editor.Full.sln, Compile.Lib.sln, and Engine.sln build just fine. So I don't know what the problem with the Travis CI build fail is. I can't do anything about it without _specific_ feedback.

ericoporto commented:
The error is not on your code, the error is because ags-palrender plugin was included in the build process but it's code is not compatible with GCC. I guess at some point you rebased or something and got that into it from master.

fernewelten commented:
> this means current master is probably broken.

Thanks for the message. The things that I currently need to get working are Bytecode concerning, which is independant of the ecosystem (Windows/Linux etc.) and doesn't rely on the ags-palrender plugin. So I'll be able to live with broken Linux builds just fine for a long time as long as I have a functioning Windows build to run my tests in.

ivan-mogilko commented:
It's not even linux build in general, but only when it has built-in plugins.

As for the plugin problem, it was fixed in ags3 branch iirc, and may get fixed in master when I finish merging latest ags3 to master. Still there could be other issues there because ags4 version has deprecated some of the plugin api, so it may take a bit to resolve everything.

gurok commented on Compiler/script/cs_parser.cpp:
```diff
@@ -1234,23 +1250,9 @@ int AGS::Parser::ParseParamlist_ParamType(AGS::Symbol param_vartype, bool &param
     if (FlagIsSet(_sym.get_flags(param_vartype), kSFLG_StructType) && (!param_is_ptr))
     {
-        cc_error("A struct cannot be passed as parameter");
+        cc_error("A non-managed struct cannot be passed as parameter");
```
I think the original error message was right here. You pass in a pointer to a managed struct. You don't pass in the struct.

fernewelten replied:
Thanks for the hint; you're right, of course.

fernewelten commented:
It seems that I was a bit to early when I assumed to be fine for the time being: The AGS Editor _compiles_, but that doesn't mean that the Editor will  _run_ (properly). To the contrary: As soon as I touch any source code window, it blows up into smoke. I've tried retrofitting my compiler to a previous change (specifically, 4191f0e1), and in this change, the Editor will compile and run properly, but f5 will yield the error: "Cannot find sources". This is probably not directly related to the Bytecode generation.

Without a working system, I can only come so far with Googletests: I've used a disassembler extensively for some weeks that can turn
```
int code[] = {
  38,    0,    6,    2,            2,   52,   48,    2,    // 7
  29,    2,    6,    2,            0,   48,    3,   34,    // 15
   3,   30,    2,   45,            2,   39,    1,    6,    // 23
   3,    1,   33,    3,           35,    1,    6,    3,    // 31
   0,    5,  -999 
};
```
into something more readable such as
```
[…]
    5:      checknull.ptr   
    6:      memread.ptr   mar
    8:      push   mar
   10:      mov    mar,   0[float 0]
   13:      memread.ptr   ax
   15:      farpush   ax
   17:      pop    mar
   19:      callobj   mar
   21:      setfuncargs   1[float 1.4013e-45]
```
But I can only test these small pieces of code by  tracing the program flow on paper. It would be about time to start crafting small pieces of dedicated code for single specific source features that is run in earnest to test whether the Bytecode does what it should do. Say, elementary mini room scripts in a `room_AfterFadeIn()`.  That would require an editor that would compile programs that can be called up with f5.

ivan-mogilko commented:
@fernewelten first of all regarding disassembler, have you tried agsutils?
https://github.com/rofl0r/agsutils/
It already has packer/extractor and assembler/dissassembler, so you could extract and run whole game scripts or parts of these if you like.

Regarding your issues with the editor itself, could you open a new ticket and describe your problems? I believe it would be much better to have separate discussion for that.

fernewelten commented:
In additon to the ~180 Googletests that cover >80% of the source lines, I've ported the compiler manually into a local branch that branches off 09725a73 (version 3.5.2), and I've run a lot of live tests to see whether the compiler generates code that the real Engine interpret as intended.

ivan-mogilko commented:
@fernewelten Could you please tell, what is the situation with this pull request and your work? Is it complete and ready for testing, or still in progress? What are your plans regarding this?

Note that in current state code cannot be merged because of some conflicts, so likely require rebase and tidying the branch.

On the other hand, if it's still not done and you think this may take longer, maybe the PR could be closed, and you'll open a new one later?

fernewelten commented:
Hi,

I don't intend to extend this PR any more except for bug fixing, porting etc.

As concerns the conflicting files, these are part of the central files of this PR, and these files have changed in more than **360** commits and **1½ years** so extensively that don't think there is any hope of GIT managing to carry over any changes automatically by now. Fundamental data structures have changed. The code is organized in a different way. 

What do we do with it?

fernewelten commented:
Tidied and rewrote the very first comment into what was added to the language, what changed in the compiler.

ivan-mogilko commented:
> I don't intend to extend this PR any more except for bug fixing, porting etc.

I see, was not sure if it's done.

> What do we do with it?

I plan to check this out again later this week. But my preliminary thought was to propose a new pull request, based on top of latest ags4 branch, with only single or few commits, because the new code completely replaces old one and hence there's little need to keep all history of changes in its branch, and frankly would be much more clear (there are 354 commits in this PR now!).
PS. Looking more at the "changed files" list, it contains files that are not even relevant to compiler. Probably these got here after merging upstream into pr branch.

ivan-mogilko commented:
Maybe you could even put your new code into the new cpp/h files?

fernewelten commented:
So AFAICS, the original compiler has gained the ability to set function defaults for non-integer variables. I've ported that functionality "by hand" to this compiler so that these changes don't get lost.

ericoporto commented:
Is it viable to emit a warning if a 0 if found as a default parameter for a float variable instead of 0.0?

fernewelten commented:
> Is it viable to emit a warning if a 0 if found as a default parameter for a float variable instead of 0.0?

There isn't any mechanism yet that will pass compiler warnings to the Editor: The only thing I can do so far is call  `cc_error()` as a sort of death message and abort the compilation run. Only the last call of `cc_error()` will stick.

If there was, it would be trivial to warn about `0` as a float default since I coded that as a special case with a dedicated `if`. The same warning would need to be given if users use the default `0` instead of `null` for pointers.

My suggestion: Let's add a third parameter to `cc_compile()` in file  Compiler\script\cs_parser.h that collects the warnings. I could then see to it that all warnings end up in this parameter. Something like
```
struct Warning
{
    int Line;
    Int Message;
};

extern int cc_compile(char const *inpl, ccCompiledScript *scrip, std::vector<Warning> &warnings);
```

AFAICS, the only place where `cc_compile()` is called is within `ccCompileText()` of file Compiler\script\cs_compiler.cpp. From there, control flow passes into the Editor. So some way could be devised to accept these warnings at this point and do something with them, e.g., display them to the user.

Unfortunately, I don't know squat about C#; so I can't help much with the C# side of the compilation.

fernewelten commented:
> "Error (line 429) 'bool' is an unmanaged struct; dynamic arrays of such types are not supported"

That's a bug: `bool` is defined to be an `enum { false = 0, true, }`, not a `struct`. So the compiler got that wrong. I'll fix this.

ericoporto commented:
You are king, this compiler is awesome! Thanks! We leave the error as is and merge, once it's merged it will be easier to work with the C# stuff, so no worries now. :D

ivan-mogilko commented:
@fernewelten, I took closer look at the contents of this branch. Besides changes to compiler it contains a number of unnecessary changes and re-added binary files (such as AGS.Native.dll), and therefore cannot be merged as-is anyway.

For the record, following changes under spoiler are the relevant ones:

<details>
  <summary>Click to expand</summary>
  (-) Compiler/fmem.cpp
  (-) Compiler/fmem.h
  (-) Compiler/script/cc_macrotable.cpp
  (-) Compiler/script/cc_macrotable.h
  (-) Compiler/script/cc_symboldef.h
  (-) Compiler/script/cc_treemap.cpp
  (-) Compiler/script/cc_treemap.h
  (-) Compiler/script/cs_prepro.cpp
  (-) Compiler/script/cs_prepro.h
  (-) Compiler/test/cc_treemap_test.cpp

  (M) Solutions/Compiler.Lib/Compiler.Lib.vcxproj
  (M) Solutions/Compiler.Lib/Compiler.Lib.Test.vcxproj
  (M) Common/script/cc_options.cpp
  (M) Common/script/cc_options.h
  (M) Compiler/script/cc_compiledscript.cpp
  (M) Compiler/script/cc_compiledscript.h
  (M) Compiler/script/cc_internallist.cpp
  (M) Compiler/script/cc_internallist.h
  (M) Compiler/script/cc_symboltable.cpp
  (M) Compiler/script/cc_symboltable.h
  (M) Compiler/script/cc_variablesymlist.h
  (M) Compiler/script/cs_compiler.cpp
  (M) Compiler/script/cs_compiler.h
  (M) Compiler/script/cs_parser.cpp
  (M) Compiler/script/cs_parser.h
  (M) Compiler/script/cs_parser_common.cpp
  (M) Compiler/script/cs_parser_common.h
  (M) Compiler/test/cs_internallist_test.cpp
  (M) Compiler/test/cs_symboltable_test.cpp
  (M) Compiler/test/cs_parser_test.cpp

  (+) Compiler/script/cs_scanner.cpp
  (+) Compiler/script/cs_scanner.h
  (+) Compiler/test/cc_bytecompat_test.cpp
  (+) Compiler/test/cs_feature_test.cpp
  (+) Compiler/test/cs_scanner_test.cpp
</details>

In addition, this change is a significant one, and may cause trouble if some feature stopped working as it may take time to fix/restore. I think it may be a good thing to keep old compiler files intact until we are sure that new one is at least as stable.


What I'd like to propose to you is this:

1) Create a new branch at the latest "ags4" branch (note, ours was updated ~1.5 months ago).
2) In that branch create Compiler/script2/ and Compiler/test2/ directories.
3) Move the new compiler and test files in these directories. In above list these are files marked as (M) and (+).
EXCEPTION 1: files Common/script/cc_options.h/cpp may be applied directly, as their changes seem universal (these are just cleanup).
EXCEPTION 2: DON'T copy MSVS project files - Compiler.Lib.vcxproj and Compiler.Lib.Test.vcxproj (see below).
4) Clone project files Compiler.Lib.vcxproj and Compiler.Lib.Test.vcxproj, name new ones Compiler2.Lib.vcxproj and Compiler2.Lib.Test.vcxproj, and replace their contents with your new compiler source.
5) Clone MSVS solution Compiler.Lib.sln, call it Compiler2.sln, and replace its projects with new ones Compiler2.Lib and Compiler2.Lib.Test.
6) In AGS.Editor.Full.sln and AGS.Editor.NoNative.sln replace Compiler.Lib with Compiler2.Lib project.
7) Create a single commit ("New compiler... etc") from these changes.
8) Open a **new pull request** from that branch.

If you have trouble with any of the above points, please tell, we may be able to help. But I hope this should not be much of a problem to do.

fernewelten commented:
@ivan-mogilko I've updated AGS 4 (checked that my version has the last commit that the AGS version has) and made a first attempt at https://github.com/adventuregamestudio/ags/pull/1101, trying to follow your instructions.

I didn't get AGS.Editor.Full to compile; I'm getting error messages that don't seem to be compiler related. I'm not sure what went wrong.

![grafik](https://user-images.githubusercontent.com/41213694/86542170-8b441480-bf13-11ea-9d8e-c9b6ca0e18f2.png)

As far as I could determine, AGS.Editor.NoNative.sln doesn't reference Compiler code.

ericoporto commented:
I noticed you created a Compiler.Lib2.vcxproj but not the Compiler.Lib2.vcxproj.filters and same for the Compiler.Lib2.Test.vcxproj and it's filters

ivan-mogilko commented:
@fernewelten  this may be a chain reaction after some error. Sometimes it happens if you do not have all necessary third-party assemblies that Editor needs. Try taking AGS 3.5.0 installation and copy all non-AGS dlls into Editor/References folder of your repository.

When rebuilding your new branch I got different errors, including one problem that I might have overlooked, I will report back when I find a solution.

fernewelten commented:
@ericoporto @ivan-mogilko 

I've fixed the header file, and I've also fixed the typos ~~Compiler.Lib2~~ Compiler2.Lib.
So now I don't get any compilation errors any more, either -- but OTOH, I run into a hundred linker errors. 

Some among them complain about a mismatch in runtime compilation flags, cf. the example below. But I've checked that Visual Studio still pulls in the same library folders as before; and I've checked that no runtime flags have changed as compared to the project files on master.
Example:
>`Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in acfonts_agsnative.obj	AGS.Native	ags\Editor\AGS.Native\Common_d.lib(roomstruct.obj)	1	`

Have the external libraries changed in some way? The ones I have are alfont-1.9.1, allegro-4.4.2, libogg-1.3.2, libtheora-1.1.1, and libvorbis-1.3.4. 
All of them in `md`, `mdd`, `mt` and `mtd` variants, i.e., static and dynamic ones.

The other linker errors seem to have the type "This has been defined before"
Example:
>`Error	LNK2005	"public: __thiscall AGS::Common::RoomBgFrame::RoomBgFrame(void)" (??0RoomBgFrame@Common@AGS@@QAE@XZ) already defined in Common_md_d.lib(roomstruct.obj)	AGS.Native	ags\Editor\AGS.Native\Common_d.lib(roomstruct.obj)	1	`

ericoporto commented:
Ahn, I think the vcxproj you pushed are switched, the one that has no 2 is referencing files from script2 directory. I can take a look tomorrow.

ivan-mogilko commented:
The Compiler.Lib/Compiler.Lib.vcxproj.filters and Compiler.Lib.Test.vcxproj.filters you commited last, they should be Compiler.Lib/Compiler2.Lib.vcxproj.filters and Compiler2.Lib.Test.vcxproj.filters, correspondingly.

ivan-mogilko commented:
The Editor builds completely for me. 
We are using allegro-4.4.3 now, but in theory Editor should be building with 4.4.2 too.
Maybe you have some leftover files in your output folders?

ivan-mogilko commented:
In both of the two error messages you posted, it mentions Common_d.lib:
> Error LNK2038 mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in acfonts_agsnative.obj AGS.Native ags\Editor\AGS.Native\Common_d.lib(roomstruct.obj) 1
> Error LNK2005 "public: __thiscall AGS::Common::RoomBgFrame::RoomBgFrame(void)" (??0RoomBgFrame@Common@AGS@@QAE@XZ) already defined in Common_md_d.lib(roomstruct.obj) AGS.Native ags\Editor\AGS.Native\Common_d.lib(roomstruct.obj) 1

But Common_d.lib should not built with AGS.Native, only Common_md_d.lib should.

ivan-mogilko commented:
Also, let's move the conversation to the new PR, since it's open now.

ivan-mogilko commented:
This PR was succeeded by #1101 .
